File: basic_dsa_forty_five.cpp
/**
 * Basic DSA - Backspace String Compare (LeetCode 844)
 * Compare two strings after applying backspace (#) processing.
 */
#include <iostream>
#include <string>

std::string build(std::string s) {
    std::string out;
    for (char c : s) {
        if (c == '#') {
            if (!out.empty()) out.pop_back();
        } else out += c;
    }
    return out;
}

bool backspaceCompare(std::string s, std::string t) {
    return build(s) == build(t);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Backspace String Compare\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "ab#c, ad#c -> " << (backspaceCompare("ab#c", "ad#c") ? "true" : "false") << "\n";
    std::cout << "ab##, c#d# -> " << (backspaceCompare("ab##", "c#d#") ? "true" : "false") << "\n";
    std::cout << "a#c, b -> " << (backspaceCompare("a#c", "b") ? "true" : "false") << "\n";
}

File: basic_dsa_forty_five.py
"""
Basic DSA - Valid Palindrome (Ignore non-alphanumeric)
Check if a string is a palindrome after converting to lowercase and removing non-alphanumeric characters.
"""


def is_valid_palindrome(s):
    """
    Two-pointer scan from both ends.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    left, right = 0, len(s) - 1

    while left < right:
        while left < right and not s[left].isalnum():
            left += 1
        while left < right and not s[right].isalnum():
            right -= 1

        if s[left].lower() != s[right].lower():
            return False

        left += 1
        right -= 1

    return True


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Valid Palindrome (Ignore non-alphanumeric)")
    print("=" * 60)

    tests = [
        ("A man, a plan, a canal: Panama", True),
        ("race a car", False),
        ("", True),
        (" ", True),
        ("0P", False),
        ("No 'x' in Nixon", True),
    ]

    for s, expected in tests:
        got = is_valid_palindrome(s)
        print(f"s={s!r} -> {got} (expected {expected})")

